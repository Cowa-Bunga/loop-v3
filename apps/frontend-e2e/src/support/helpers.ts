export const testIdSelector = (testId: string) => `[data-testid="${testId}"]`

export const COOKIE_NAME = 'cowabunga-user-cookie'
