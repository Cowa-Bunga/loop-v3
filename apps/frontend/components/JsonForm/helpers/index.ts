const makeInputId = (id: string, suffix?: string) => {
  const parts = id.split('/')

  if (suffix) {
    return `${parts[parts.length - 1]}-${suffix}`
  }

  return parts[parts.length - 1]
}

export { makeInputId }
