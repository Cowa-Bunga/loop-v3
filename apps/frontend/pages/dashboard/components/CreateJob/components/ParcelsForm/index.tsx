import { JsonForm } from '@components'
import type { JsonSchema, Layout } from '@jsonforms/core'
import { Grid, IconButton, Typography } from '@mui/material'
import { GridAddIcon } from '@mui/x-data-grid'

interface IProps {
  schema: JsonSchema
  ui: Layout
  model: any
  onChange: (index: number, data: any) => void
  addParcel: () => void
  parcels: object[]
}

const ParcelsForm = ({
  ui,
  model,
  schema,
  onChange,
  addParcel,
  parcels
}: IProps) => {
  return (
    <Grid
      container
      style={{ padding: '0 20px ' }}
      justifyContent="space-between"
      columns={16}
    >
      <Grid item>
        <Typography variant="h5">Parcels</Typography>
      </Grid>
      <Grid item>
        <IconButton aria-label="Add parcel" color="primary" onClick={addParcel}>
          <GridAddIcon />
        </IconButton>
      </Grid>

      {parcels.map((parcel, index) => (
        <JsonForm
          key={index}
          schema={schema}
          ui={ui}
          model={model}
          onChange={(e) => onChange(index, e)}
        />
      ))}
    </Grid>
  )
}

export default ParcelsForm
