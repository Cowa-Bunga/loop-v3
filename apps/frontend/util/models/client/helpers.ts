import { defaultsDeep } from 'lodash'
import { MODEL_CLIENT_EMPTY } from './model'

export const modelClientList = (
  data: { [s: string]: unknown } | ArrayLike<unknown>
): IClient[] => {
  return Object.entries(data)
    .map(modelClientData)
    .map((el: never) => el['1'])
    .sort((a: IClient, b: IClient) => a.name.localeCompare(b.name))
}

export const modelClientData = (data: unknown): IClient => {
  return defaultsDeep(data, MODEL_CLIENT_EMPTY)
}
