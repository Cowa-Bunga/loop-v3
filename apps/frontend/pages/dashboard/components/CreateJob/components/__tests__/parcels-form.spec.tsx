import ParcelsForm from '../ParcelsForm'
import { fireEvent, render } from '@testing-library/react'
import { renderWithProviders } from '@test/helpers'
import * as axe from 'axe-core'

const mockSchema = {
  schema: {
    type: 'object',
    properties: {
      qrCode: {
        type: 'string'
      },
      size: {
        type: 'string',
        enum: ['small', 'medium', 'large'] // enums should be localized keys
      }
    }
  },

  ui: {
    type: 'HorizontalLayout',
    elements: [
      {
        type: 'TextInputControl',
        scope: '#/properties/qrCode',
        i18n: 'create_task_form.parcels.qrCode'
      },
      {
        type: 'DropdownControl',
        scope: '#/properties/size',
        i18n: 'create_task_form.parcels.size'
      }
    ]
  },

  data: {
    qrCode: '',
    size: ''
  }
}

jest.mock('react-i18next', () => ({
  ...jest.requireActual('react-i18next'),
  getI18n: () => ({
    resolvedLanguage: 'en'
  })
}))

jest.mock('@hooks', () => ({
  ...jest.requireActual('@hooks'),
  useTranslation: () => ({
    t: jest.fn().mockImplementation((key) => key)
  })
}))

describe('ParcelsForm', () => {
  it('should render with no parcels', () => {
    const { baseElement } = renderWithProviders(
      <ParcelsForm
        schema={mockSchema.schema}
        model={mockSchema.data}
        ui={mockSchema.ui}
        addParcel={jest.fn()}
        onChange={jest.fn()}
        parcels={[]}
      />
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('should render with  parcels', () => {
    const { baseElement } = render(
      <ParcelsForm
        schema={mockSchema.schema}
        model={mockSchema.data}
        ui={mockSchema.ui}
        addParcel={jest.fn()}
        onChange={jest.fn()}
        parcels={[
          {
            size: 'small',
            qrCode: '123'
          }
        ]}
      />
    )
    expect(baseElement).toMatchSnapshot()
  })

  it('should call addParcel when IconButton is clicked', () => {
    const mockAddParcel = jest.fn()
    const { baseElement } = renderWithProviders(
      <ParcelsForm
        schema={mockSchema.schema}
        model={mockSchema.data}
        ui={mockSchema.ui}
        addParcel={mockAddParcel}
        onChange={jest.fn()}
        parcels={[
          {
            size: 'small',
            qrCode: '123'
          }
        ]}
      />
    )

    const addButton = baseElement.querySelector(
      "button[aria-label='Add parcel']"
    )

    fireEvent.click(addButton)

    expect(mockAddParcel).toHaveBeenCalled()
  })

  it('Accessibility check', async () => {
    const { container } = renderWithProviders(
      <ParcelsForm
        schema={mockSchema.schema}
        model={mockSchema.data}
        ui={mockSchema.ui}
        addParcel={jest.fn()}
        onChange={jest.fn()}
        parcels={[]}
      />
    )
    const results = await axe.run(container)
    expect(results.violations.length).toBe(0)
  })
})
