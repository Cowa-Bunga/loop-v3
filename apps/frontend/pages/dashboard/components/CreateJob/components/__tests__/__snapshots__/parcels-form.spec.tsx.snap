// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParcelsForm should render with  parcels 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      style="padding: 0px 20px;"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Parcels
        </h5>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          aria-label="Add parcel"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`ParcelsForm should render with no parcels 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container css-1xzqqmo-MuiGrid-root"
      style="padding: 0px 20px;"
    >
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
        >
          Parcels
        </h5>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <button
          aria-label="Add parcel"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
